### ISSUE TITLE
Documentation - misunderstandings, (in)completeness, ...

### ISSUE THREAD

**Issue by dumblob** [vlang/v#4520]
I noticed https://github.com/vlang/v/blob/master/doc/docs.md became the de-facto official introduction/doc for *V*, but it doesn't seem to cover whole *V* (e.g. typedefs, variant types, ...), but more importantly specifies a bit different behavior of optional types than I understood.

I think the following is incorrect:

>You can also propagate errors:
>
> ```v
> resp := http.get(url)?
> println(resp.body)
> ```
>`http.get` returns `?http.Response`. It was called with `?`, so the error is propagated to the calling function (which must return an optional) or in case of `main` leads to a panic. Basically the code above is a shorter version of
>```v
>resp := http.get(url) or {
>    panic(err)
>}
>println(resp.body)
>```
>V does not have a way to force unwrap an optional (like Rust's `unwrap()` or Swift's `!`). You have to use `or { panic(err) }` instead.

And should rather be (notice `return error( err )` instead of `panic( err )`):

>You can also propagate errors:
>
> ```v
> resp := http.get(url)?
> println(resp.body)
> ```
>`http.get` returns `?http.Response`. It was called with `?`, so the error is propagated to the calling function (which must return an optional) or in case of `main` leads to a panic. Basically the code above is a shorter version of
>```v
>resp := http.get(url) or {
>    return error(err)
>}
>println(resp.body)
>```
>V does not have a way to force unwrap an optional (like Rust's `unwrap()` or Swift's `!`). You have to use `or { return error(err) }` instead.

@medvednikov ?

**Comment by medvednikov** [comment]
> And should rather be (notice return error( err ) instead of panic( err )):

no, panic() is correct, unwrap() stops program's execution in Rust

**Comment by dumblob** [comment]
Hm, IMHO that's a complete **game changer** as the question mark syntax `resp := http.get(url)?` strongly encourages people to write unsafe code :cry::

1. because it's easier to write than the safe and well-defined and more efficient `resp := http.get( url ) or { return error( err ) }`

1. it won't run any deferred stuff in functions higher in the stack (thus making a lot of assumptions about imported modules etc. utterly wrong leading to widespread unsafe usage)

2. code, which doesn't allow for interruption efficiently and readably in the code (co interrupt a `panic()`, one needs to very verbosely write a separate `defer{...}` line for each such call and each such interruption is not efficient and get's very complex (to define and remember all the subtle edge cases) and completely **impossible** to make run fast once multithreading etc. will be implemented

Basically this semantics of `?` goes totally against *V* goals as the simplicity it provides makes the language magnitudes more unsafe and complex to reason about. Note also, that Rust deals with the above issues differently and thus it's to *V* as apples to oranges.

I'd strongly suggest rethinking this and at least changing the semantics to `or { return error(err) }` or better [scratch `panic()` completely](https://github.com/vlang/v/issues/2030 ) to make the language simpler and more transparent (and safer and faster).

@joe-conigliaro @medvednikov @elimisteve @danieldaeschle @radare @spytheman

**Comment by danieldaeschle** [comment]
`or { return error(err) }` has another meaning than you think.
Return in an or block returns a value of the function. I think you mean `or { error(err) }`.

But error is a value and `panic` exits the program.

**Comment by dumblob** [comment]
>`or { return error(err) }` has another behaviour than you think.
Return in an or block returns a value for the function. I think you mean `or { error(err) }`.

No, I really meant `or { return error(err) }` to propagate the error (thus making the compiler shout if the return value is not an optional.

>But error is a value and panic exits the program.

Returning an error in `main()` also exits the program. The propagation as written above ensures the rest.

>You think writing or { panic(err) } everywhere is better than a question mark?

No, the point is, that `panic()` won't be needed at all (or just for debugging/prototyping - e.g. only in the non-`-prod` build).

>Imo a programmer should know what a behaviour of a programming language is.

Yes. But a programmer won't read millions of SLOC of modules she uses.

**Comment by spytheman** [comment]
In my opinion, you should make a PR @dumblob.

**Comment by spytheman** [comment]
`x := opt()?` *is* a shortcut for `x := opt() or {return error( err ) }` in functions != main, where the containing function returns an optional too.

**Comment by spytheman** [comment]
For example, consider this function:
```v
import os
fn all_lines_from_file(f string) ?[]string{
    content := os.read_file()?
    return content.split('\n')
}
```
This is equivalent to:
```v
import os
fn all_lines_from_file(f string) ?[]string{
    content := os.read_file() or { return error(err) }
    return content.split('\n')
}
```
In case the file can not be read, or does not exist, os.read_file will fail, and its error will get propagated as the error of the custom function all_lines_from_file.

The caller of all_lines_from_file can handle the error in an or block, or if the caller is also a function that returns an optional, can also do all_lines_from_file()?, which will propagate the error further up the stack.

If the caller is the main function, *then* calling `all_lines_from_file()?` will not propagate the error further (fn main() does not return an option), instead the call will be equivalent to `panic(err)`, on case of failure.

**Comment by medvednikov** [comment]
> Hm, IMHO that's a complete game changer as the question mark syntax resp := http.get(url)? strongly encourages people to write unsafe code 

I think you misunderstand what `?` does, @spytheman gave a good summary.

Also why would you remove panics, what would be used instead?

**Comment by dumblob** [comment]
Thanks @spytheman for explanation - what you've explained is exactly the behavior I'd expect. Yay!

In that case the documentation is indeed misleading. I'll consider a PR this time :wink:.

>Also why would you remove panics, what would be used instead?

One would always use `return error( ... )` (or `?`) instead. The *V* language specification would then make a special case for `main()` which would also work with `return error( ... )`, but it'd specify: *`stderr` is used to print the content of `err` with one subsequent LF and a non-zero status is returned* (I wouldn't specify *a status `1` will be returned* to have space for future releases where one could want to distinguish certain error classes).

**Comment by medvednikov** [comment]
But why? There are countless situations where an application needs to quit immediately.

**Comment by dumblob** [comment]
>But why? There are countless situations where an application needs to quit immediately.

Of course and I think `return error( ... )` satisfies them (it's as easy as adding `?` to the return type of a function and to the given line in this function and you're done).

The main reason is, that the code at the AST leafs must not decide what to do with the whole app (imagine a subtle unimportant module will panic, but the app does much more important and dangerous things with different side effects and resources like fs/DB/network locks, but suddenly gets unsafely interrupted by such an unimportant module).

Propagating each such error (i.e. making it explicit) forces the root nodes of AST handle these consciousnessly without beind dependent on deep-in modules buried decisions someone else made for a generic use case of an unimportant module.

As I said, `panic()` is good for debugging etc., but not for a safe production multithreaded code. If `panic()` shall stay in production builds of *V* code, then I'd at least suggest having all standard modules/libraries which go bundled with *V* completely `panic()` free to ensure safety at least in the canonical *V* implementation.

**Comment by elimisteve** [comment]
> As I said, `panic()` is good for debugging etc., but not for a safe production multithreaded code. 

Right, which is why `?` works the way that it does, and why `panic` should rarely be used.

What should happen when we divide by zero? I can imagine V requiring that the user conditionally checks that the denominator != 0 before doing the division, but it'd probably all end up being too awkward, so `panic`king seems best AFAICT.

> imagine a subtle unimportant module will panic, but the app does much more important and dangerous things with different side effects and resources like fs/DB/network locks, but suddenly gets unsafely interrupted by such an unimportant module

@medvednikov It will be possible to `recover` from a `panic`, as in Go, yes?

**Comment by dumblob** [comment]
>What should happen when we divide by zero? I can imagine V requiring that the user conditionally checks that the denominator != 0 before doing the division, but it'd probably all end up being too awkward, so panicking seems best AFAICT.

Why `panic()`king? Why not `return error( 'division by 0' )`? Btw. there are edge cases when division by 0 makes sense, so **by default** *V* shouldn't probably do anything about this and leave the result as is (i.e. as computed by the underlying hardware). Checking for infinity/undef/flags just extremely slows down arithmetics anyway.

>@medvednikov It will be possible to recover from a panic, as in Go, yes?

The point is, recovering is:
1) verbose
2) error prone (usually people won't think of it - look at how people write code in Python which is notoriously known for modules throwing hundreds of exceptions, but still people don't care and don't catch them - that's **reality** :cry:)
3) slow (can't be made fast)
4) unnecessary (it can be substituted easily with `return error( ... )` which doesn't have any of these flaws; second, it violates the fundamental *V* principle *there is only one way of doing things*)

**Comment by lutherwenxu** [comment]
I assume that this issue has been resolved by your PR in #4551.
Please reopen if it's still not resolved.

**Comment by dumblob** [comment]
In the doc there is still missing e.g. typedefs, variant types, ..., but I can't make a PR, because I don't know the exact semantics :cry:. Someone from the *V* designers core team has to write it.
